### PR TITLE
[rush] Add `--ignore-hooks` command line flag to all rush actions

### PR DESCRIPTION
**Pull request by dmichon-msft**
Adds a new command line flag `--ignore-hooks` to all rush actions that does exactly what it says. Skips any configured `eventHooks` in `rush.json` during execution of the action. Example usage is when the actions are to perform telemetry for local dev installs and the commands are being invoked on a CI machine.

Edited: per feedback, switched from only `rush install` and `rush update` to all rush actions.

### CODE REVIEW

**Comment by D4N14L** [comment]
Should we create an equivalent arg on build/rebuild since we have hooks there as well?

**Comment by dmichon-msft** [comment]
Probably, though for the time being I'm mostly concerned with install/update and the code for build is somewhere else altogether.

File: apps/rush-lib/src/cli/actions/BaseInstallAction.ts [hunk 1]
```diff
@@ -77,6 +78,10 @@ export abstract class BaseInstallAction extends BaseRushAction {
       description: `Overrides the default maximum number of install attempts.`,
       defaultValue: RushConstants.defaultMaxInstallAttempts
     });
+    this._ignoreHooksParameter = this.defineFlagParameter({
+      parameterLongName: '--ignore-hooks',
+      description: `Overrides execution of "preInstall" and "postInstall" event hooks. Meant for use on CI machines.`
```
Comment (D4N14L) [inline apps/rush-lib/src/cli/actions/BaseInstallAction.ts:83]: > Meant for use on CI machines.

Not sure if this bit is necessary. It's true in our case, but not necessarily for others

**Review by D4N14L — APPROVED** [review]
I'm okay with this, but should also get input from @iclanton or @pgonzal since this is a new install param and we try to limit introduction of new params

File: common/changes/@microsoft/rush/dmichon-add-ignore-hooks_2020-10-14-23-17.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Add --ignore-hooks flag",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/dmichon-add-ignore-hooks_2020-10-14-23-17.json:5]:
```suggestion
      "comment": "Add a --ignore-hooks flag to the rush install and rush update commands.",
```

File: apps/rush-lib/src/cli/actions/BaseInstallAction.ts [hunk 3]
```diff
@@ -97,7 +102,11 @@ export abstract class BaseInstallAction extends BaseRushAction {
       StandardScriptUpdater.validate(this.rushConfiguration);
     }
 
-    this.eventHooksManager.handle(Event.preRushInstall, this.parser.isDebug);
+    if (!this._ignoreHooksParameter.value) {
+      this.eventHooksManager.handle(Event.preRushInstall, this.parser.isDebug);
+    } else {
+      console.log(`The --ignore-hooks flag was specified, skipping preInstall hooks`);
```
Comment (iclanton) [inline apps/rush-lib/src/cli/actions/BaseInstallAction.ts:108]: We should probably only log these messages if any hooks are actually defined.

Comment (dmichon-msft) [reply apps/rush-lib/src/cli/actions/BaseInstallAction.ts:108]: Does it make sense to delegate this behavior to the EventHooksManager, then?

Comment (iclanton) [reply apps/rush-lib/src/cli/actions/BaseInstallAction.ts:108]: I think so.

File: apps/rush-lib/src/logic/EventHooksManager.ts [hunk 4]
```diff
@@ -21,13 +21,18 @@ export class EventHooksManager {
...
 
-  public handle(event: Event, isDebug: boolean): void {
+  public handle(event: Event, isDebug: boolean, ignoreHooks: boolean): void {
     if (!this._eventHooks) {
       return;
     }
 
     const scripts: string[] = this._eventHooks.get(event);
     if (scripts.length > 0) {
+      console.log(`Skipping event hooks for ${Event[event]} since --ignore-hooks was specified`);
```
Comment (D4N14L) [inline apps/rush-lib/src/logic/EventHooksManager.ts:31]: I believe this should go under the `if` statement below

File: apps/rush-lib/src/cli/actions/BaseInstallAction.ts (hunk 3 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/actions/BaseInstallAction.ts]: Moved.

File: apps/rush-lib/src/logic/EventHooksManager.ts (hunk 4 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/logic/EventHooksManager.ts:31]: ... oops

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts [hunk 5]
```diff
@@ -219,6 +220,10 @@ export class BulkScriptAction extends BaseScriptAction {
           'but not any projects that directly or indirectly depend on the changed package.'
       });
     }
+    this._ignoreHooksParameter = this.defineFlagParameter({
+      parameterLongName: '--ignore-hooks',
+      description: `Overrides execution of event hooks. Meant for use on CI machines.`
```
Comment (octogonz) [inline apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts]:
```suggestion
      description: `Skips invocation of the "eventHooks" scripts that are defined in rush.json. This maybe useful for automated jobs.`
```

File: common/changes/@microsoft/rush/dmichon-add-ignore-hooks_2020-10-14-23-17.json (hunk 2 above)
Comment (octogonz) [reply common/changes/@microsoft/rush/dmichon-add-ignore-hooks_2020-10-14-23-17.json]: `"comment": "Add an --ignore-hooks flag to the \"rush install\" and \"rush update\" commands."`

**Review by octogonz — APPROVED** [review]
AWS

File: apps/rush-lib/src/cli/actions/BaseInstallAction.ts (hunk 1 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/actions/BaseInstallAction.ts]: Updated.

File: common/changes/@microsoft/rush/dmichon-add-ignore-hooks_2020-10-14-23-17.json (hunk 2 above)
Comment (dmichon-msft) [reply common/changes/@microsoft/rush/dmichon-add-ignore-hooks_2020-10-14-23-17.json]: The underlying text has been updated to match the behavior following previous feedback to move the logic to the EventHooksManager and thereby to all rush actions

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts (hunk 5 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts]: Applied the first half. I prefer the current version of the second half (no statement of where you want it, just to be careful using it).

**Comment by dmichon-msft** [comment]
@octogonz , @iclanton , have altered text per your suggestions if you want to give this a final look before merge.

**Review by iclanton — APPROVED** [review]

**Pull request merged**